cookbook/security/guard-authentication.rst

```diff
@@ -0,0 +1,577 @@
+    :ref:`AuthenticationException <guard-customize-error>`)
+    authentication will fail.
+
+**checkCredentials($credentials, UserInterface $user)**
+    If ``getUser()`` returns a User object, this method is called. Your job is to
+    verify if the credentials are correct. For a login for, this is where you would
+    check that the password is correct for the user. To pass authentication, return
+    ``true``. If you return *anything* else
+    (or throw an :ref:`AuthenticationException <guard-customize-error>`),
+    authentication will fail.
+
+**onAuthenticationSuccess(Request $request, TokenInterface $token, $providerKey)**
+    This is called after successful authentication and your job is to either
+    return a :class:`Symfony\\Component\\HttpFoundation\\Response` object
+    that will be sent to the client or ``null`` to continue the request
+    (e.g. allow the route/controller to be called like normal). Since this
+    is an API where each request authenticates itself, you want to return
+    ``nul``.
+
+**onAuthenticationFailure(Request $request, AuthenticationException $exception)**
+    This is called if authentication fails. Your job
+    is to return the :class:`Symfony\\Component\\HttpFoundation\\Response`
+    object that should be sent to the client. The ``$exception`` will tell you
+    *what* went wrong during authentication.
+
+**start**
+    This is called if the client accesses a URI/resource that requires authentication,
+    but no authentication details were sent (i.e. you returned ``null`` from
+    ``getCredentialsFromRequest()``). Your job is to return a
+    :class:`Symfony\\Component\\HttpFoundation\\Response` object that helps
+    the user authenticate (e.g. a 401 response that says "token is missing!").
+
+**supportsRememberMe**
+    Since this is a stateless API, you do not want to support "remember me"
+    functionality.
+
+.. _guard-customize-error:
+
+Customizing Error Messages
+--------------------------
+
+When ``onAuthenticationFailure()`` is called, it is passed an ``AuthenticationException``
+that describes *how* authentication failed via its ``$e->getMessageKey()`` (and
+``$e->getMessageData()``) method. The message will be different based on *where*
+authentication fails (i.e. ``getUser()`` versus ``checkCredentials()``).
+
+But, you can easily return a custom message by throwing a
+:class:`Symfony\\Component\\Security\\Core\\Exception\\CustomUserMessageAuthenticationException`.
+You can throw this from ``getCredentials()``, ``getUser()`` or ``checkCredentials()``
+to cause a failure::
+
+    // src/AppBundle/Security/TokenAuthenticator.php
+    // ...
+    
+    use Symfony\Component\Security\Core\Exception\CustomUserMessageAuthenticationException;
+    
+    class TokenAuthenticator extends AbstractGuardAuthenticator
+    {
+        // ...
+        
+        public function getCredentials(Request $request)
+        {
+            // ...
+            
+            if ($token == 'MickyMouse') {
```

Comment:
Apart rom being a copyrighted name by the toughest copyright enforcer company in the world, I'm not sure I like the "Micky Mouse" example.